### PR TITLE
Fetch Model Information

### LINKED ISSUE #128: [Backend] Store additional model information via ModelAppConfig
AC: ModelAppConfig should be extended to include model metadata, like input types, output types, parameters, and constraints.

### PR DESCRIPTION
**Pull request by ShriramG24**
Closes #128.

Note: I'm not sure I like the way I'm handling errors with `fetchModelInfo` by setting the return type to `Promise<ModelInfo | null>` (check out `super-res-inference-service.ts` for example). Let me know if there's a better way to go about it. I was thinking of reusing the SuccessResponse/ErrorResponse classes but their data fields are of type `object`, which isn't restrictive enough compared to the `ModelInfo` type.

image

### CODE REVIEW

**Comment by atharvakale343** [comment]
Is the current way of defining a model details page too restrictive?

**Comment by ShriramG24** [comment]
> Is the current way of defining a model details page too restrictive? Should model devs be able to control this page? Perhaps the backend API can be provided. And the devs can choose to include some/all/none of this information, and include more information without an API.

Honestly, I think we'd be better off leaving it to the model devs. Although, I think we should still maintain some uniformity between the inspect pages for different models. Maybe we can factor out the right portion (containing the box) as a common component and leave the entire left side as a separate component for the model devs to design.

For this PR, I guess I can leave in the API logic and create a separate issue for whatever we decide to do regarding the inspect model page.

**Comment by atharvakale343** [comment]
Sounds good.

File: src/main/handlers/registration.ts [hunk 1]
```diff
@@ -52,9 +59,25 @@ const getModelAppStatus = async (
...
+
+const getModelInfo = async (
+  _event: any,
+  arg: GetModelInfoArgs,
+): Promise<ModelInfo> => {
+  const server = await ModelServer.getServerByModelUid(arg.modelUid);
+  if (!server) {
+    throw new Error(`Server not found for model ${arg.modelUid}.`);
+  }
+  const manager = new JobManager(getServiceByModelUid(arg.modelUid));
```
Comment (atharvakale343) [inline src/main/handlers/registration.ts:71]: JobManager never changed?

File: src/main/model-apps/super-res/super-res-inference-service.ts [hunk 2]
```diff
@@ -77,6 +80,64 @@ class SuperResInferenceService implements InferenceService {
...
+        {
+          name: 'Minimum Scale',
+          description: 'The minimum scale factor is 1.',
+        },
+        {
+          name: 'Maximum Scale',
+          description: 'The maximum scale factor is 4.',
+        },
+      ],
+    } as ModelInfo;
```
Comment (atharvakale343) [inline src/main/model-apps/super-res/super-res-inference-service.ts:139]: you shouldn't be needing `as ModelInfo` if this object satisfies the type. If you're just trying to enforce it, you can use `} satisfies ModelInfo`.

File: src/shared/models.ts [hunk 3]
```diff
@@ -15,3 +15,16 @@ export enum ModelAppStatus {
   Error = 'Error',
   Unregistered = 'Unregistered',
 }
+
+export type ModelInfo = {
```
Comment (atharvakale343) [inline src/shared/models.ts:19]: there is a lot of duplication between `ModelAppConfig` defined in config.ts and `ModelInfo`. We should only put fields in one or the other. Perhaps ModelInfo here can exclude the first 5 fields.

**Review by atharvakale343 — CHANGES_REQUESTED** [review]

**Comment by ShriramG24** [comment]
Will make a separate PR. Most of the changes in this one aren't relevant anymore.

**Pull request closed without merging**